### PR TITLE
feat: Add `Change data directory location` patch

### PR DESCRIPTION
**Pull request by KrystianQur**
Added a patch to automate changing app files location from data/0 to android/data for root-less devices

### CODE REVIEW

**Comment by oSumAtrIX** [comment]
What are these locations and what do they have to do with root?

**Comment by KrystianQur** [comment]
Application on data/0 can not be accessed without root, by changing everything in the application to "getFilesExternalDir" the application fully is accessible at android/data instead of data/0 without needing root for access

**Comment by Ushie** [comment]
What's the benefit of this? the expected usage?

**Comment by KrystianQur** [comment]
For some applications/games it is useful to modify the files from here, which are hidden without the patch

**Review by oSumAtrIX — COMMENTED** [review]
If an app uses the external directory already, will that cause issues?

**Comment by indrastorms** [comment]
android/data is also inaccessible in android 12+ without adb/shizuku

**Comment by KrystianQur** [comment]
@indrastorms I have android 14 and can normally access android/data without adb/shizuku

**Comment by KrystianQur** [comment]
@oSumAtrIX No

**Comment by oSumAtrIX** [comment]
And what if the app writes to a file called a in the external directory and the internal?

**Comment by KrystianQur** [comment]
Too nothing, the patch searches for "getDir" methods and replaces with these external ones

**Comment by indrastorms** [comment]
> @indrastorms I have android 14 and can normally access android/data without adb/shizuku

Not possible maybe your oem doesn't restrict it.
https://developer.android.com/about/versions/11/privacy/storage#other-apps-data

**Comment by oSumAtrIX** [comment]
This is offtopic. The patch says non-root and that's the case here

**Comment by KrystianQur** [comment]
Android 14 got new security ways which ZArchiver uses to access android/data as I know

File: src/main/kotlin/app/revanced/patches/all/directory/ChangeDataDirectoryLocationPatch.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
+import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
+import com.android.tools.smali.dexlib2.iface.reference.MethodReference
+import com.android.tools.smali.dexlib2.iface.instruction.formats.Instruction35c
+import com.android.tools.smali.dexlib2.immutable.reference.ImmutableMethodReference
+import com.android.tools.smali.dexlib2.util.MethodUtil
+
+@Patch(
+    name = "Change data directory location",
+    description = "Changes data directory in the application from data/0 to android/data for root-less devices.",
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/all/directory/ChangeDataDirectoryLocationPatch.kt:20]: Please mention the full path here:

```suggestion
    description = "Changes the data directory in the application from data/0 to android/data for root-less devices.",
```

**Comment by oSumAtrIX** [comment]
> Too nothing, the patch searches for "getDir" methods and replaces with these external ones

Yes, so if it returns the external directory instead which is already used by the app for other purposes it could cause conflicts like in the example i mentioned.

**Comment by KrystianQur** [comment]
> > Too nothing, the patch searches for "getDir" methods and replaces with these external ones
> 
> Yes, so if it returns the external directory instead which is already used by the app for other purposes it could cause conflicts like in the example i mentioned.

Yeah maybe, but I tested in few app's, and it was working fine.

File: src/main/kotlin/app/revanced/patches/all/directory/ChangeDataDirectoryLocationPatch.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
+import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
+import com.android.tools.smali.dexlib2.iface.reference.MethodReference
+import com.android.tools.smali.dexlib2.iface.instruction.formats.Instruction35c
+import com.android.tools.smali.dexlib2.immutable.reference.ImmutableMethodReference
+import com.android.tools.smali.dexlib2.util.MethodUtil
+
+@Patch(
+    name = "Change data directory location",
+    description = "Changes the data directory in the application from /data/user/0 to /sdcard/android/data for root-less devices.",
```
[the remark this answers is not in the document]
Comment (KrystianQur) [reply src/main/kotlin/app/revanced/patches/all/directory/ChangeDataDirectoryLocationPatch.kt:20]: Yeah, I guess this is good idea

**Pull request merged**